### ISSUE TITLE
Improved fallback to null/absent value

### ISSUE THREAD

**Issue by ben-manes** [jhalterman/failsafe#177]
Using v1, I had code that used a `null` fallback to indicate a soft error such as quota limits, rather than hard fail the outer execution. In v2 a `null` value is not allowed, so instead I'm converting to `Optional.<T>empty()`. That requires the type hint, which is a bit verbose.

Perhaps there could be a synchronous version of `Failsafe.get` that wraps the result, e.g. often expressed as a `Try<T>`? Then instead of providing a `Fallback`, I could use `failsafe.tryGet(...)` instead and handle it accordingly.

**Comment by jhalterman** [comment]
We could certainly make a null fallback allowed if it makes sense. I don't think there was any specific reason not to allow it, come to think of it. Have an example use case?

**Comment by ben-manes** [comment]
We were hitting rate limiting with Mapbox route projections, since we do it on the server-side whereas they expect it to be on the client, so a server's IP is much more active than a mobile's. Since projections are non-critical (e.g. ETA for delivery), we can reprocess soft failures later without blocking other work on that process.

The v1 code is below. I typically use `@Nullable <T>` within a class and `Optional<T>` for public APIs, which I think is a good balance as `null` for internal details should be very obvious compared to an opaque call to an external class.

```java
@Inject
public MapboxClient(@MapboxAccessToken String accessToken) {
  this.rateLimiter = RateLimiter.create(PERMITS_PER_SECOND);
  this.accessToken = accessToken;
  this.failsafe = Failsafe
      .with(new RetryPolicy()
          .withDelay(1, TimeUnit.SECONDS)
          .withMaxRetries(3)
      .retryWhen(null))
      .withFallback(() -> null)
      .onFailedAttempt(t -> logger.warn("Route projection failed", t));
}

public Optional<ProjectedRoute> projectRoute(Geolocation origin, Geolocation destination) {
  DirectionsResponse response =  failsafe.get(() -> getDirections(origin, destination));
  if ((response == null) || isEmpty(response.getWaypoints())) {
    return Optional.empty();
  }
  ...
}

private @Nullable DirectionsResponse getDirections(Geolocation origin, Geolocation destination) {
  try {
    rateLimiter.acquire();
    MapboxDirections client = ...
    Response<DirectionsResponse> response = client.executeCall();
    if (response.isSuccessful()) {
      return response.body();
    }
    logger.warn("[{}] {}", response.code(), response.errorBody().string());
    return null;
  } catch (IOException | ServicesException e) {
    logger.error("Failed to retreive driving directions from Mapbox", e);
    return null;
  }
}
```

**Comment by ben-manes** [comment]
The only problem with `null` value in `Fallback` is that it would look like `Fallback.of((DirectionsResponse) null)` in the `Failsafe.with` chain, since typing is now necessary and can be a little verbose. The `tryGet` api would be an alternative option to consider (both could be good changes).

**Comment by jhalterman** [comment]
Picking this up again after a while... :)

Allowing null in Fallback is fine, but yea, a cast would be needed.

Did you have a particular `Try` implementation in mind? I'm hesitant to add a bare bones implementation when users may prefer something like Vavr's Try impl.

**Comment by ben-manes** [comment]
I use Vavr's `Try`, but I am not partial to that library. It was mostly an idea of an API, nothing prescriptive as I migrated this usage.

**Comment by jhalterman** [comment]
Cool. With the support for `Fallback.of((Foo)null)`, will close this for now.

**Comment by jhalterman** [comment]
This was released in 2.1.
